### ISSUE TITLE
bogus Diagnostic for static const functions

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#8289]
|  |  |
| --- | --- |
| Bugzilla Link | [7917](https://llvm.org/bz7917) |
| Version | trunk |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
When it encounters a const static function Clang will not only report this but also a 'conflicting types' error that actually does not exist in the source. It looks like Clang just ignores the 'const' in the declaration which then leads to inconsistency with the definition. (Hope I did not mix up declaration and definition...)

example:
--------

$ cat test.cpp 
```c++
class Foo {
  static void bar() const;
};
void Foo::bar() const {}
```

$ clang++ test.cpp 
```
test.cpp:2:15: error: type qualifier is not allowed on this function
  static void bar() const;
              ^
test.cpp:5:11: error: conflicting types for 'bar'
void Foo::bar() const {}
          ^
test.cpp:2:15: note: previous declaration is here
  static void bar() const;
              ^
2 errors generated.
```

**Comment by philnik777** [comment]
Still in trunk: https://godbolt.org/z/4dh6Eh5Gh

**Comment by shafik** [comment]
static member function cannot be qualified `const` see [class.static.mfct p2](https://eel.is/c++draft/class.static.mfct#2):

> [Note [2](https://eel.is/c++draft/class.static.mfct#note-2): A static member function does not have a this pointer ([[expr.prim.this]](https://eel.is/c++draft/expr.prim.this))[.](https://eel.is/c++draft/class.static.mfct#2.sentence-1) A static member function cannot be qualified with const, volatile, or virtual ([[dcl.fct]](https://eel.is/c++draft/dcl.fct))[.](https://eel.is/c++draft/class.static.mfct#2.sentence-2) — end note]

Note gcc also diagnoses this: https://godbolt.org/z/9a7rfejhj

**Comment by shafik** [comment]
It looks like we remove the cv-qualifiers from the function declaration after we diagnose that it can not be labeled `static` here:

https://github.com/llvm/llvm-project/blob/8b6f09e257b521947b50761737290e5bf31c80f3/clang/lib/Sema/SemaType.cpp#L5854-L5861

I have not tested if any test break if we remove this line but I have tested that the `conflicting types` diagnostic does go away.

**Comment by shafik** [comment]
In after thought, this makes sense since we should drop top level cv-qualifiers here even though it is invalid.

**Comment by shafik** [comment]
I am not sure but if I am reading the code correctly I think in order to handle this correctly we need to be able see the previous declaration but debugging the code it does not look like we have that information available at this point.

@zygoloid any tips here?

**Comment by zygoloid** [comment]
It seems to me that we have three options here:

1) recover from the first error by removing the `const`, leading to the diagnostics we see here and (for example) errors if `this` is used from the function body
2) recover from the first error by removing the `static`, which would fix the subsequent error here but might lead to other errors (eg, if the function is used without an object)
3) don't recover and instead mark the function declaration as invalid; ensure that we don't complain if we attempt to define a function that is marked invalid

None of these is really ideal, but given that there are two likely interpretations of the original code, perhaps we shouldn't be guessing and should prefer 3 over 1 or 2? Or perhaps we should prefer option 2 since it seems less likely to lead to follow-on errors.
